src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/IO/TdsWriteStream.cs

```diff
@@ -0,0 +1,308 @@
+        public override bool CanSeek => false;
+
+        /// <summary>
+        /// Returns true since we can always write to the stream.
+        /// </summary>
+        public override bool CanWrite => _underlyingStream != null;
+
+        /// <inheritdoc />
+        public override long Length
+            => throw new NotSupportedException();
+
+        /// <inheritdoc />
+        public TdsStreamPacketType? PacketHeaderType { get; set; }
+
+        /// <inheritdoc />
+        public override long Position
+        {
+            get => throw new NotSupportedException();
+            set => throw new NotSupportedException();
+        }
+
+        #endregion
+
+        #region Public Methods
+
+        /// <inheritdoc />
+        public override async ValueTask DisposeAsync()
+        {
+            await _underlyingStream.DisposeAsync();
+            _underlyingStream = null;
+            _writeBuffer = null;
+        }
+
+        /// <inheritdoc />
+        public override void Flush()
+            => FlushPacketAsync(true, isAsync: false, CancellationToken.None).Wait();
+
+        /// <inheritdoc />
+        public override async Task FlushAsync(CancellationToken ct)
+            => await FlushPacketAsync(true, true, ct).ConfigureAwait(false);
+
+        /// <inheritdoc />
+        public void SetPacketSize(int bufferSize)
+        {
+            _writeBuffer = new byte[bufferSize];
+            InitializeHeaderBytesToZero();
+        }
+
+        /// <inheritdoc />
+        public void ReplaceUnderlyingStream(Stream stream) => _underlyingStream = stream;
+
+        /// <inheritdoc />
+        public override int Read(byte[] buffer, int offset, int count) => throw new NotSupportedException();
+
+        /// <inheritdoc />
+        public override long Seek(long offset, SeekOrigin origin) => throw new NotSupportedException();
+
+        /// <inheritdoc />
+        public override void SetLength(long value) => throw new NotSupportedException();
+
+        /// <inheritdoc />
+        public override void Write(byte[] buffer, int offset, int count)
+            => Write(buffer.AsSpan(offset, count));
+
+
+        /// <inheritdoc/>
+        public override void Write(ReadOnlySpan<byte> buffer)
+        {
+            int len = buffer.Length;
+            // The buffer may not have enough space. Write what we can and then flush the buffer with a soft flush.
+            while (len > 0)
+            {
+                if (len > _writeBuffer.Length - _writeBufferEnd)
+                {
+                    // Only a part of the length fits in the buffer.
+                    int bytesToWrite = _writeBuffer.Length - _writeBufferEnd;
+                    buffer[..bytesToWrite].CopyTo(_writeBuffer.AsSpan(_writeBufferEnd));
+                    _writeBufferEnd += bytesToWrite;
+                    len -= bytesToWrite;
+                    FlushPacketAsync(false, isAsync: false, CancellationToken.None).ConfigureAwait(false);
```

Comment:
Not awaited.
Same comment as above for WriteByte.

Reply:
No need to await here no ? The FlusPacketAsync will complete synchronously due to the flag.

Reply:
Also I cant await in a sync method.